contrib/parseq-batching/src/main/java/com/linkedin/parseq/batching/BatchingStrategy.java

```diff
@@ -0,0 +1,245 @@
+ * </ol>
+ *
+ * @author Jaroslaw Odzga (jodzga@linkedin.com)
+ *
+ * @param <G> Type of a Group
+ * @param <K> Type of a Key
+ * @param <T> Type of a Value
+ */
+public abstract class BatchingStrategy<G, K, T> {
+
+  private final ConcurrentHashMap<Long, BatchBuilder<K, T>> _batches =
+      new ConcurrentHashMap<>();
+
+  /**
+   * This method returns Task that returns value for a single key in allowing this strategy to batch operations.
+   * @param desc description of the task
+   * @param key key
+   * @return Task that returns value for a single key in allowing this strategy to batch operations
+   */
+  public Task<T> batchable(final String desc, final K key) {
+    return Task.async(desc, ctx -> {
+      final SettablePromise<T> result = Promises.settable();
+      Long planId = ctx.getPlanId();
+      BatchBuilder<K, T> builder = _batches.get(planId);
+      if (builder == null) {
+        builder = Batch.builder();
+        BatchBuilder<K, T> existingBuilder = _batches.putIfAbsent(planId, builder);
+        if (existingBuilder != null) {
+          builder = existingBuilder;
+        }
+      }
+      builder.add(key, ctx.getShallowTraceBuilder(), result);
+      return result;
+    });
+  }
+
+  private Task<?> taskForBatch(final G group, final Batch<K, T> batch) {
+    return Task.async(getBatchName(batch, group), ctx -> {
+      final SettablePromise<T> result = Promises.settable();
+      final PromiseListener<T> countDownListener =
+          new CountDownPromiseListener<T>(batch.size(), result, null);
+
+      //one of the batchable tasks becomes a parent of batch task
+      //all other tasks become potential parents
+      final TraceBuilder traceBuilder = ctx.getTraceBuilder();
+      Relationship rel = Relationship.PARENT_OF;
+      for (BatchEntry<T> entry : batch.values()) {
+        for (ShallowTraceBuilder shallowTraceBuilder: entry.getShallowTraceBuilders()) {
+          traceBuilder.addRelationship(rel, shallowTraceBuilder, ctx.getShallowTraceBuilder());
+          rel = Relationship.POTENTIAL_PARENT_OF;
+        }
+        entry.getPromise().addListener(countDownListener);
+      }
+
+      try {
+        if (batch.size() == 1) {
+          Entry<K, BatchEntry<T>> entry = batch.entires().iterator().next();
+          executeSingleton(group, entry.getKey(), entry.getValue());
+        } else {
+          executeBatch(group, batch);
+        }
+      } catch (Throwable t) {
+        batch.failAllRemaining(t);
+      }
+
+      return result;
+    });
+  }
+
+  private Collection<Task<?>> taskForBatches(Collection<Entry<G, Batch<K, T>>> batches) {
+    return batches.stream().map(entry -> taskForBatch(entry.getKey(), entry.getValue())).collect(Collectors.toList());
+  }
+
+  void handleBatch(final PlanContext planContext) {
+    final BatchBuilder<K, T> batchBuilder = _batches.remove(planContext.getId());
```

Comment:
So the BatchBuilder is added to _batches if at least one batchable task is created. And then here it gets removed, all Tasks in the same plan will have the same plan context, so this is guaranteed to work. Do I have that right?